src/gui/windows/predictor/predictor_btb_dock.cpp

```diff
@@ -6,158 +6,158 @@ DockPredictorBTB::DockPredictorBTB(QWidget *parent) : Super(parent) {
-        // Get cell item, or create new one if needed
-        QTableWidgetItem *item { btb->item(row_index, column_index) };
-        if (item == nullptr) {
-            item = new QTableWidgetItem();
-            btb->setItem(row_index, column_index, item);
-        }
-
-        // Set color
+        QTableWidgetItem *item { get_btb_cell_item(row_index, column_index) };
         item->setBackground(QBrush(color));
     }
-};
+}
 
 void DockPredictorBTB::setup(
     const machine::BranchPredictor *branch_predictor,
     const machine::Core *core) {
-    connect(
-        branch_predictor, &machine::BranchPredictor::update_btb_row_done, this,
-        &DockPredictorBTB::update_row);
-    connect(
-        branch_predictor, &machine::BranchPredictor::requested_bht_target_address, this,
-        &DockPredictorBTB::highligh_row_after_prediction);
-    connect(core, &machine::Core::step_started, this, &DockPredictorBTB::reset_colors);
+
+    clear();
 
     number_of_bits = init_number_of_bits(branch_predictor->get_number_of_btb_bits());
     const bool is_predictor_enabled { branch_predictor->get_enabled() };
 
     if (is_predictor_enabled) {
         btb->setRowCount(qPow(2, number_of_bits));
         btb->setDisabled(false);
+        clear_btb();
+
+        connect(
+            branch_predictor, &machine::BranchPredictor::btb_row_updated,
+            this, &DockPredictorBTB::update_btb_row);
+        connect(
+            branch_predictor, &machine::BranchPredictor::prediction_done,
+            this, &DockPredictorBTB::highligh_row_after_prediction);
+        connect(
+            branch_predictor, &machine::BranchPredictor::update_done,
+            this, &DockPredictorBTB::highligh_row_after_update);
+        connect(
+            core, &machine::Core::step_started,
+            this, &DockPredictorBTB::reset_colors);
+        
     } else {
         btb->setRowCount(0);
         btb->setDisabled(true);
     }
-    init_table();
-    btb->resizeRowsToContents();
-    set_table_color(Q_COLOR_DEFAULT);
-};
-
-void DockPredictorBTB::update_row(
-    uint16_t index,
-    machine::Address instruction_address,
-    machine::Address target_address) {
-    if (index >= btb->rowCount()) {
-        WARN("BTB dock update received invalid row index: %u", index);
+}
+
+void DockPredictorBTB::update_btb_row(
+    uint16_t row_index,
+    machine::BranchTargetBufferEntry btb_entry
+) {
+    if (row_index >= btb->rowCount()) {
+        WARN("BTB dock update received invalid row index: %u", row_index);
         return;
     }
 
-    set_row_color(index, Q_COLOR_UPDATE);
+    QTableWidgetItem *item;
 
-    QTableWidgetItem *item_index { btb->item(index, DOCK_BTB_COL_INDEX) };
-    QTableWidgetItem *item_instr_addr { btb->item(index, DOCK_BTB_COL_INSTR_ADDR) };
-    QTableWidgetItem *item_target_addr { btb->item(index, DOCK_BTB_COL_TARGET_ADDR) };
+    if (btb_entry.entry_valid) {
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_INSTR_ADDR);
```

Comment:
nit: Duplicated code. The initialization is same in both branches. Should be hoisted before the if and merged with the initializer.

Reply:
Never mind. I misunderstood the code.